extensions/opc/src/opc.cpp

```diff
@@ -0,0 +1,461 @@
+    elem->referenceTypeId = UA_NODEID_NUMERIC(0, ids[i]);
+    elem->targetName = UA_QUALIFIEDNAME_ALLOC(0, tokens[i].c_str());
+  }
+
+  UA_TranslateBrowsePathsToNodeIdsRequest request;
+  UA_TranslateBrowsePathsToNodeIdsRequest_init(&request);
+  request.browsePaths = &browsePath;
+  request.browsePathsSize = 1;
+
+  UA_TranslateBrowsePathsToNodeIdsResponse response = UA_Client_Service_translateBrowsePathsToNodeIds(client, request);
+
+  if(response.resultsSize < 1) {
+    logger->log_warn("No node id in response for %s", path.c_str());
+    return UA_STATUSCODE_BADNODATAAVAILABLE;
+  }
+
+  bool foundData = false;
+
+  for(size_t i = 0; i < response.resultsSize; ++i) {
+    UA_BrowsePathResult res = response.results[i];
+    for(size_t j = 0; j < res.targetsSize; ++j) {
+      foundData = true;
+      UA_NodeId resultId;
+      UA_NodeId_copy(&res.targets[j].targetId.nodeId, &resultId);
+      foundNodeIDs.push_back(resultId);
+      std::string namespaceUri((char*)res.targets[j].targetId.namespaceUri.data, res.targets[j].targetId.namespaceUri.length);
+    }
+  }
+
+  UA_BrowsePath_deleteMembers(&browsePath);
+  UA_TranslateBrowsePathsToNodeIdsResponse_deleteMembers(&response);
+
+  if(foundData) {
+    logger->log_debug("Found %lu nodes for path %s", foundNodeIDs.size(), path.c_str());
+    return UA_STATUSCODE_GOOD;
+  } else {
+    logger->log_warn("No node id found for path %s", path.c_str());
+    return UA_STATUSCODE_BADNODATAAVAILABLE;
+  }
+}
+
+nodeData getNodeData(opc::ClientPtr& clientPtr, const UA_ReferenceDescription *ref, const std::string& basePath) {
+  if(ref->nodeClass == UA_NODECLASS_VARIABLE)
+  {
+    opc::nodeData nodedata;
+    std::string browsename(reinterpret_cast<const char*>(ref->browseName.name.data), ref->browseName.name.length);
+
+    if(ref->nodeId.nodeId.identifierType == UA_NODEIDTYPE_STRING) {
+      std::string nodeidstr(reinterpret_cast<const char*>(ref->nodeId.nodeId.identifier.string.data),
+                            ref->nodeId.nodeId.identifier.string.length);
+      nodedata.attributes["NodeID"] = nodeidstr;
+      nodedata.attributes["NodeID type"] = "string";
+    } else if(ref->nodeId.nodeId.identifierType == UA_NODEIDTYPE_BYTESTRING) {
+      std::string nodeidstr(reinterpret_cast<const char*>(ref->nodeId.nodeId.identifier.byteString.data), ref->nodeId.nodeId.identifier.byteString.length);
+      nodedata.attributes["NodeID"] = nodeidstr;
+      nodedata.attributes["NodeID type"] = "bytestring";
+    } else if (ref->nodeId.nodeId.identifierType == UA_NODEIDTYPE_NUMERIC) {
+      nodedata.attributes["NodeID"] = std::to_string(ref->nodeId.nodeId.identifier.numeric);
+      nodedata.attributes["NodeID type"] = "numeric";
+    }
+    nodedata.attributes["Browsename"] = browsename;
+    nodedata.attributes["Full path"] = basePath + "/" + browsename;
+    nodedata.dataTypeID = UA_TYPES_COUNT;
+    UA_Variant* var = UA_Variant_new();
```

Comment:
This will leak if `UA_Client_readValueAttribute` fails.

Reply:
Added a delete call there.